### PR TITLE
FEATURE: ContentGraphAdapter for write side access

### LINKED ISSUE #4973: TASK: Introduce (internal) low level content graph api for constraint checks and write side
-------

**Marc Henry's part / requirements:**

In order to get the NodeIdentity/NodeAdress change finished (see https://github.com/neos/neos-development-collection/pull/4868) we must rewrite the write sides projection access. Because of the dilemma here: https://github.com/neos/neos-development-collection/pull/4868#discussion_r1531194274
(The dilemma is that the Node read model has to know its WorkspaceName somehow)

We discussed that the high level api of the content graph will eventually look like `findNodeAggregateById(WorkspaceName $ws, $id);` and also `getSubgraph` will accept the `WorkspaceName` instead of a content stream id.

So why can we not just always work with `WorkspaceName`'s also on the write side?
Because the first thing we do is resolve the current content stream we want to operate on via `requireContentStreamId` which might return, if in a rebase closure, a globally temporary mock instead of the actual one for the `WorkspaceName`.

Thus we need to have a low level api where we not only pass the `ContentStreamId` but also the `WorkspaceName`.

The `WorkspaceName` is currently not yet part of any node / node aggregate read model, but will be with the introduction of the NodeIdentity/NodeAdress ON the node. So this part has to be forward implemented without being used by the NodeFactory ... _that_ change is my part. 

During this the subtask https://github.com/neos/neos-development-collection/issues/4942 also needs to be solved, as the 2 commands dont contain `WorkspaceName`'s yet and we would need them to pass to the lower level methods.

---------

Feel free to edit this issue and add your insights.

-----

Slack threads !!!:
- https://neos-project.slack.com/archives/C04PYL8H3/p1711793524319239?thread_ts=1711791103.956609&cid=C04PYL8H3
- https://neos-project.slack.com/archives/C04PYL8H3/p1711349027838899
- https://neos-project.slack.com/archives/C04PYL8H3/p1711111681225429

### PR DESCRIPTION
**Pull request by kitsunet**
The write side no longer uses any regular finders
for checks but only accesses projection data via
the new `ContentGraphAdapterInterface`.

Fixes: #4973

### CODE REVIEW

**Comment by kitsunet** [comment]
Few notes here for starters:

* I omitted visibility constraints everywhere as we can just assume "see everything" given that that was hardcoded before.
* I added csId + workspaceName to all the methods that need it, a tuple _might_ really make this nicer to read, but it works
* I originally thought we expose the factory and not the implementation and you can grab an implementation bound to csId + workspace _but_ there is a bunch of methods in there that do not rely on it and so it felt strange as you wouldn't have either information if you need methods that do not need it.
* it's huuuuuuge, I haven't gone over and tried to reduce it by merging similar functionality, but maybe that's an option. 
* The interface and implementation should probably move to another place, I will snatch that from bastis PR.

**Comment by kitsunet** [comment]
Also I realized too late but why is e.g. the workspace projection in CR Core? it should be in the storage adapter implementations (so DBAL, postgres) etc. ? Same for the contentStreams...

**Comment by bwaidelich** [comment]
> Also I realized too late but why is e.g. the workspace projection in CR Core? it should be in the storage adapter implementations

Is that related?
But for the record: Those projections work with MySQL _and_ PostgreSQL – If it _was_ part of the storage adapter we would need to copy all of the implementations to both adapters.
But I agree that it should be in separate packages

**Comment by kitsunet** [comment]
More note.... I guess the adapter needs a bunch more dependencies to actually be able to provide Node(s) results, we'll see how that works out, they are obviously all available when building it, but I would rather want to avoid adding them...

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+use Neos\ContentRepository\Core\NodeType\NodeTypeName;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Node;
+use Neos\ContentRepository\Core\Projection\ContentGraph\NodeAggregate;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateId;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeName;
+use Neos\ContentRepository\Core\SharedModel\Workspace\ContentStreamId;
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+
+/**
+ *
```
Comment (bwaidelich) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php]: This needs some thorough documentation and an `@internal` annotation

**Comment by kitsunet** [comment]
> > Also I realized too late but why is e.g. the workspace projection in CR Core? it should be in the storage adapter implementations
> 
> Is that related? 

I tried to get rid of ALL access to finders within the handlers, so now I also took the worksapce and contentstream related methods into the adapter, which doesn't work because the DBAL adapter implementation then doesn't know about those :D

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    )
+    {
+    }
+
+    public function rootNodeAggregateWithTypeExists(ContentStreamId $contentStreamId, NodeTypeName $nodeTypeName): bool
+    {
+        // TODO: Implement rootNodeAggregateWithTypeExists() method.
+    }
+
+    public function findParentNodeAggregates(ContentStreamId $contentStreamId, WorkspaceName $workspaceName, NodeAggregateId $childNodeAggregateId): iterable
```
Comment (bwaidelich) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php]: Not directly related but we should really introduce a `NodeAggregates` DTO

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapterFactory.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+namespace Neos\ContentGraph\DoctrineDbalAdapter\Domain\Repository;
+
+use Doctrine\DBAL\Connection;
+use Neos\ContentGraph\DoctrineDbalAdapter\DoctrineDbalContentGraphProjectionFactory;
+use Neos\ContentRepository\Core\Feature\ContentGraphAdapterFactoryInterface;
+use Neos\ContentRepository\Core\Feature\ContentGraphAdapterInterface;
+use Neos\ContentRepository\Core\SharedModel\ContentRepository\ContentRepositoryId;
+
+/**
+ * This factory
```
Comment (bwaidelich) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapterFactory.php]: ...?  :)

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php [hunk 4]
```diff
@@ -387,11 +387,12 @@ protected function areNodeTypeConstraintsImposedByGrandparentValid(
      */
     protected function requireProjectedNodeAggregate(
         ContentStreamId $contentStreamId,
-        NodeAggregateId $nodeAggregateId,
-        ContentRepository $contentRepository
+        NodeAggregateId $nodeAggregateId
     ): NodeAggregate {
-        $nodeAggregate = $contentRepository->getContentGraph()->findNodeAggregateById(
+        $workspace = $this->getContentGraphAdapter()->findWorkspaceByCurrentContentStreamId($contentStreamId);
+        $nodeAggregate = $this->getContentGraphAdapter()->findNodeAggregateById(
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php]: IMO one of the goals of the adapter should be that we _don't_ have to fetch the workspace in order to retrieve node aggregate details.
The `ContentStreamIdOverride` should happen _in_ the adapter (or rather: we should probably fetch a different instance if the override is active)

**Review by bwaidelich — COMMENTED** [review]
Thanks so much for your efforts!
I just did a first quick pass as preparation for our weekly.

Apart from the nitpicky comments, I think the adapter should always act on a workspace/contentstream

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php (hunk 1 above)
Comment (kitsunet) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php]: Yep, documentation is missing _everywhere_ just wanted to get this out here for now. Important though, so thanks!

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php (hunk 2 above)
Comment (kitsunet) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapter.php]: Yes, can do maybe

File: Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapterFactory.php (hunk 3 above)
Comment (kitsunet) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraphAdapterFactory.php]: ...needs more documentation 😆 see above

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php (hunk 4 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php]: Mmm, that is also an option obviously, but then you have zero outside control, that is, you could _in theory_ get nodes with one workspace and then the other a bit later? Lets discuss.

File: Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    public function getExpectedVersionOfContentStream(): ExpectedVersion;
+    /*
+     * WORKSPACES
+     */
+
+    public function findWorkspaceByCurrentContentStreamId(
+        ContentStreamId $contentStreamId
+    ): ?Workspace;
+
+    public function withWorkspaceName(WorkspaceName $workspaceName): ContentGraphAdapterInterface;
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php]: nope, we have the adapter for this now, this was an inbetween thing.

File: Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+    public function getWorkspaceName(): WorkspaceName;
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php]: These getters for the WorkspaceName and ContentStreamId are super useful if you follow the trail throughout the nested functions in traits and such. I think it makes sense.

File: Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    public function findStateForContentStream(): ?ContentStreamState;
+
+    public function findVersionForContentStream(): MaybeVersion;
+
+    public function getExpectedVersionOfContentStream(): ExpectedVersion;
+    /*
+     * WORKSPACES
+     */
+
+    public function findWorkspaceByCurrentContentStreamId(
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php]: This also is unused now and will be removed in the next update

File: Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+    public function findSuceedingSiblingNodesInSubgraph(
+        DimensionSpacePoint $coveredDimensionSpacePoint,
+        NodeAggregateId $startingSiblingNodeAggregateId
+    ): Nodes;
+
+    /*
+     * CONTENT STREAMS
+     */
+
+    public function hasContentStream(): bool;
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Feature/ContentGraphAdapterInterface.php:164]: as outlined by the ContentGraphAdapterProviderInterface I think that one should just throw and not even offer us an adapter if there is no content stream?

**Comment by kitsunet** [comment]
> Now come the hard(er) parts in a way. The nitty gritty. Some of the _missing_ methods (see linter errors) were removed by me because phpstorm didn't detect usage. The interesting question is which of those use cases do we consider valid "read" use cases that should use the content graph and which should use the new ContentGraphAdapter. I think the two controllers are probably _real_ use cases and should be re-instated to work. the ContentCacheFlusher for example I would rather solve with the ContentGraphAdapter as well as the SiteServiceInternals...

All resolved for now, but I will probably poke a few other corners. Also now need to deduplicate the queries.

File: Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapterFactoryBuilder.php [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+use Neos\ContentRepository\Core\Factory\ProjectionFactoryDependencies;
+use Neos\ContentRepository\Core\Feature\ContentGraphAdapterFactoryBuilderInterface;
+use Neos\ContentRepository\Core\Infrastructure\DbalClientInterface;
+
+/**
+ * Builder to combine injected dependencies and ProjectionFActoryDependencies into a ContentGraphAdapterFactory
+ * @internal
+ */
+class ContentGraphAdapterFactoryBuilder implements ContentGraphAdapterFactoryBuilderInterface
```
Comment (kitsunet) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapterFactoryBuilder.php:16]: I wouldn't mind getting around having this, and am happy for suggestions, I don't really see it though. An ugly option would be to to have something along the lines of:

`ContentGraphAdapterFactory::injectProjectionDependencies(ProjectionFactoryDependencies $projectionFactoryDependencies)`

and the `ContentGraphAdapterFactory::__construct` just gets the adapter dependencies injected (eg. DBALClient), but then all methods in the ContentGraphAdapterFactory need to throw a "TooEarly" Exception if the projection dependencies were not injected yet.

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php [hunk 10]
```diff
@@ -224,6 +224,8 @@ public function eventNumberIs(int $eventNumber, string $eventType, TableNode $pa
                 $actual = DimensionSpacePointSet::fromArray($actualValue);
                 Assert::assertTrue($expected->equals($actual), 'Actual Dimension Space Point set "' . json_encode($actualValue) . '" does not match expected Dimension Space Point set "' . $assertionTableRow['Expected'] . '"');
             } else {
+//                die();
```
Comment (kitsunet) [inline Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php]: weird, I am not even sure why I woudl want to do something here^^ Removing

File: Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapterFactoryBuilder.php (hunk 9 above)
Comment (bwaidelich) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapterFactoryBuilder.php:16]: AFAIS we only need it to pass in `ContentRepositoryId`, `NodeTypeManager` and `PropertyConverter` – Maybe we can make these explicit dependencies somehow

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php (hunk 10 above)
Comment (bwaidelich) [reply Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php]: ?

File: Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php [hunk 11]
```diff
@@ -224,6 +224,8 @@ public function eventNumberIs(int $eventNumber, string $eventType, TableNode $pa
...
+//                exit;
```
Comment (bwaidelich) [inline Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php]: Ah you mean, you want to remove those changes again? I agree :)
```suggestion
```

File: Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapterFactoryBuilder.php (hunk 9 above)
Comment (bwaidelich) [reply Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapterFactoryBuilder.php:16]: ..or even make the `NodeFactory` an explicit dependency (via some `NodeFactoryInterface`) but I don't know... Let's discuss

**Comment by kitsunet** [comment]
In our weekly 2024-04-19 we talked about how this adapter is maybe not worth the trouble and if we can solve the ugly contentStreamIdOverride differently without this. And it is indeed difficult to make assumptions for after introduction of node identity, but it is very clear to me that at a higher level commands and API will use workspaceName the queries (at least for the DBAL Adapter) will need a contentStreamId to work, therefore this translation will have to happen somewhere, IMHO the changes for NodeIdentity will just push this deeper into the read side making it even more annoying (e.g. due to`ContentSubgraphWithRuntimeCaches`) to override for the few caess that we have. And those are valid cases as we apply commands with $some workspaceName to a contentStream that has (not yet got) this workspaceName.

This together with the (from the POV of the write side) convoluted API to read stuff lets me think that this change could have merit even IF we cannot at this point avoid using read models (IMHO this could be a later improvement though) and much easier to just change the write side interface for this.

Convoluted because we apply filters, visibility and a lot of other stuff.

I am still not in love with the triple factory thing, but the separation of APIs seems very sensible still.

File: Neos.ContentRepositoryRegistry.DoctrineDbalClient/Configuration/Settings.yaml [hunk 12]
```diff
@@ -7,3 +7,5 @@ Neos:
           # catchUpHooks for content cache flushing
           'Neos.ContentRepository:ContentGraph':
             factoryObjectName: Neos\ContentGraph\DoctrineDbalAdapter\DoctrineDbalContentGraphProjectionFactory
+        contentGraphAdapterFactory:
+          factoryObjectName: Neos\ContentGraph\DoctrineDbalAdapter\ContentGraphAdapterFactoryBuilder
```
Comment (mhsdesign) [inline Neos.ContentRepositoryRegistry.DoctrineDbalClient/Configuration/Settings.yaml:11]: optional followup would be to simplify this config

File: Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php [hunk 13]
```diff
@@ -281,4 +285,16 @@ private function buildClock(ContentRepositoryId $contentRepositoryIdentifier, ar
...
+    private function buildContentGraphAdapterProvider(ContentRepositoryId $contentRepositoryIdentifier, array $contentRepositorySettings): ContentGraphAdapterFactoryBuilderInterface
+    {
+        isset($contentRepositorySettings['contentGraphAdapterFactory']['factoryObjectName']) || throw InvalidConfigurationException::fromMessage('Content repository "%s" does not have contentGraphAdapterFactory.factoryObjectName configured.', $contentRepositoryIdentifier->value);
+        $contentGraphAdapterFactoryBuilder = $this->objectManager->get($contentRepositorySettings['contentGraphAdapterFactory']['factoryObjectName']);
+        if (!$contentGraphAdapterFactoryBuilder instanceof ContentGraphAdapterFactoryBuilderInterface) {
+            throw InvalidConfigurationException::fromMessage('contentGraphAdapterFactory.factoryObjectName for content repository "%s" is not an instance of %s but %s.', $contentRepositoryIdentifier->value, ContentGraphAdapterFactoryBuilderInterface::class, get_debug_type($contentGraphAdapterFactoryBuilder));
+        }
+        // TODO: Do we want to add options here? Would then have to be a setter I guess....
```
Comment (mhsdesign) [inline Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php:297]: todo over here ;)

File: Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapter.php
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapter.php:1]: We should refine this naming

File: Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapter.php [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+        }
+
+        $this->contentStreamId = ContentStreamId::fromString($contentStreamIdString);
+
+        return $this->contentStreamId;
+    }
+
+    public function getWorkspace(): Workspace
+    {
+        $query = $this->dbalConnection->prepare('SELECT * FROM cr_default_p_workspace WHERE workspacename LIKE :workspaceName');
```
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapter.php:485]: illegal select, and the `Workspace` tdo is too big

File: Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapter.php [hunk 15]
```diff
@@ -0,0 +1,507 @@
...
+    }
+
+    public function getContentStreamId(): ContentStreamId
+    {
+        if ($this->contentStreamId !== null) {
+            return $this->contentStreamId;
+        }
+
+        // TODO: This table is not allowed here...
+        $query = $this->dbalConnection->prepare('SELECT currentcontentstreamid FROM cr_default_p_workspace WHERE workspacename LIKE :workspaceName');
```
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphAdapter.php:471]: todo illegal

**Comment by bwaidelich** [comment]
This is to be replaced by #5028 and I suggest to close the PR (we can still keep the branch for reference if needed)

**Pull request closed without merging**